Firestore/core/test/firebase/firestore/util/async_queue_test.cc

```diff
@@ -0,0 +1,362 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/util/async_queue.h"
+
+#include <stdlib.h>
+#include <chrono>  // NOLINT(build/c++11)
+#include <future>  // NOLINT(build/c++11)
+#include <string>
+#include <thread>  // NOLINT(build/c++11)
+
+#include <gtest/gtest.h>
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+namespace {
+
+namespace chr = std::chrono;
+
+chr::time_point<std::chrono::system_clock, chr::milliseconds> now() {
+  return chr::time_point_cast<AsyncQueue::Milliseconds>(
+      chr::system_clock::now());
+}
+
+const auto kTimeout = std::chrono::seconds(5);
+
+// Waits for the future to become ready and returns whether it timed out.
+bool WaitForFuture(const std::future<void>& future,
+                   const chr::milliseconds timeout = kTimeout) {
+  return future.wait_for(timeout) == std::future_status::ready;
+}
+
+// Unfortunately, the future returned from std::async blocks in its destructor
+// until the async call is finished. If the function called from std::async is
+// buggy and hangs forever, the future's destructor will also hang forever. To
+// avoid all tests freezing, the only thing to do is to abort (which skips
+// destructors).
+void Abort() {
+  ADD_FAILURE();
+  std::abort();
+}
+
+// Calls std::abort if the future times out.
+void AbortOnTimeout(const std::future<void>& future) {
+  if (!WaitForFuture(future, kTimeout)) {
+    Abort();
+  }
+}
+
+// The macro calls AbortOnTimeout, but preserves stack trace.
+#define ABORT_ON_TIMEOUT(future)                            \
+  do {                                                      \
+    SCOPED_TRACE("Async operation timed out, aborting..."); \
+    AbortOnTimeout(future);                                 \
+  } while (0)
+
+class ScheduleTest : public ::testing::Test {
+ public:
+  ScheduleTest() : start_time{now()} {
+  }
+
+  using ScheduleT = Schedule<int>;
+
+  int TryPop() {
+    int read = -1;
+    schedule.PopIfDue(&read);
+    return read;
+  }
+
+  ScheduleT schedule;
+  ScheduleT::TimePoint start_time;
+  int out = 0;
+};
+
+}  // namespace
+
+// Schedule tests
+
+TEST_F(ScheduleTest, PopIfDue_Immediate) {
+  EXPECT_FALSE(schedule.PopIfDue(nullptr));
+
+  // Push values in a deliberately non-sorted order.
+  schedule.Push(3, start_time);
+  schedule.Push(1, start_time);
+  schedule.Push(2, start_time);
+  EXPECT_FALSE(schedule.empty());
+  EXPECT_EQ(schedule.size(), 3u);
+
+  EXPECT_EQ(TryPop(), 3);
+  EXPECT_EQ(TryPop(), 1);
+  EXPECT_EQ(TryPop(), 2);
+  EXPECT_FALSE(schedule.PopIfDue(nullptr));
+  EXPECT_TRUE(schedule.empty());
+  EXPECT_EQ(schedule.size(), 0u);
+}
+
+TEST_F(ScheduleTest, PopIfDue_Delayed) {
+  schedule.Push(1, start_time + chr::milliseconds(5));
+  schedule.Push(2, start_time + chr::milliseconds(3));
+  schedule.Push(3, start_time + chr::milliseconds(1));
+
+  EXPECT_FALSE(schedule.PopIfDue(&out));
```

Comment:
Technically, this suffers from a timing issue. If there's a context switch anywhere between lines 114-116 inclusive that causes 5ms+ delay, then this could legitimately return true resulting in a flakey test. (This issue appears in a few places below too.)

I think this could be addressed by injecting a fake clock. (Possibly as a template parameter to Scheulde and AsyncQueue.)

Reply:
sgtm